rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,344 @@
+- Feature Name: versioning-types
+- Start Date: 2022-06-28
+- RFC PR: [blockprotocol/blockprotocol#408](https://github.com/blockprotocol/blockprotocol/pull/408)
+- Block Protocol Discussion: [blockprotocol/blockprotocol#408](https://github.com/blockprotocol/blockprotocol/discussions/408)
+
+# Summary
+
+[summary]: #summary
+
+The Type System outlined in [RFC 0352](https://github.com/blockprotocol/blockprotocol/blob/am/rfc-graph-type-system/rfcs/text/0352-graph-type-system.md) introduced classes of types that are used to describe the data used and produced by blocks and embedding applications. The types are possibly (ideally often) publicly accessible, and their hierarchy allows for them to be composed together into new types. Due to this, changes in a type can affect the structure of another one. This RFC defines how types are versioned, to mitigate potential problems that would arise without versioning.
+
+# Motivation
+
+[motivation]: #motivation
+
+The Type System is intended to allow for **reusable and shared** descriptions of data. As such, a version of a block should be able to tie itself to a _specific_ description of the structure of the data it requires as if types are able to be updated independently of blocks then this would cause problems. Additionally, if someone (Alice) creates an entity type that refers to someone else's (Bob's) "address" property type, we do not want Bob to be able to change what Alice meant in her entity type by arbitrarily and implicitly updating it.
+
+Being able to refer to specific, immutable versions of types should mitigate or solve the majority of the issues stemming from this.
+
+This RFC proposes that every Type (i.e. Data Type, Property Type, Entity Type, or Link Type) should be associated with a unique version number. This version number will be part of the URI that is used to uniquely identify the specific instance of the Type, to access the Type's schema, and to refer to the Type from other Types.
+
+We also outline a method to compare compatibility between types, and use this in the [Rationale and Alternatives](#rationale-and-alternatives) section to explain how this replaces the need for some other versioning scheme such as semantic versioning.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Type Versions
+
+The _version number_ of a type is an incrementing positive integer that increases by 1 with each new version, starting from 1.
+
+## Type URIs
+
+The _base URI_ of a type is a _unique_ identifier for the type irrespective of its version.
+
+The _versioned URI_ of a type is made up of:
+
+```json
+"${base_uri}/v/${version_number}"
```

Comment:
I also am torn between the options, I was discussing it with @thehabbos007 and we couldn't decide easily. I had a number of reasons, not that I found them particularly convincing
- For self-hosting types, it's a lot easier in most implementations to just have it as part of the route, rather than needing to understand how to interpret query parameters and route more dynamically based on that
- One less special-ish character
- NPM uses `/v/`, I hope that has reasoning behind it that we can benefit from
- I think `/v/` will play nicer with file-systems if people really want to do so

Reply:
I don't have anything more to add on this, really. Maybe see if you can drum up any other opinions as part of implementation

Reply:
We should keep in mind that people may want more variations of types than simply the versions. For example, folks might want to use query params to request/serve a localized translation _of_ a specific version of a type for their own i18n.

If we use `/v/` paths for versions we should be clear why, and still consider how we would support layering of additional parameters as well.

Reply:
That's a pretty good argument (in my head) for not talking about query params at the moment at all, and using `/v/` to keep it simple. We _could_ do the i18n thing later on as you say, and at that point query params would be a good fit. I'd probably prefer if we never went down that road because this isn't just about UX but data structure and I think trying to translate keys in actual JSON data is perhaps not a good idea from a technical standpoint. But anyway, we aren't ruling it out as part of this RFC.